webapp/cube/views.py

```diff
@@ -133,6 +144,149 @@ def cube_microcerts():
     )
 
 
+@login_required
+def get_microcerts():
+
+    # backend_true requires query string,
+    # therefore we need context
+    # also this avoids loading contracts api where not required
+    is_test_backend = (
+        current_app.config["CONTRACTS_TEST_API_URL"]
+        if flask.request.args.get("test_backend", "false")
+        else current_app.config["CONTRACTS_LIVE_API_URL"]
+    )
+    sso_user = user_info(flask.session)
+    if not sso_user:
+        if flask.request.path != "/cube/microcerts":
+            return flask.redirect(
+                "/cube/microcerts?test_backend=true"
+                if is_test_backend
+                else "/cube/microcerts"
+            )
+
+        return flask.render_template(
+            "cube/index-no-login.html",
+            is_test_backend=is_test_backend,
+        )
+
+    ua_contracts_api = UAContractsAPI(
+        session=ua_contracts_session,
+        authentication_token=(flask.session.get("authentication_token")),
+        api_url=is_test_backend,
+    )
+
+    edx_url = (
+        f"{edx_api.base_url}/auth/login/tpa-saml/"
+        "?auth_entry=login&idp=ubuntuone&next="
+    )
+
+    edx_user = edx_api.get_user(sso_user["email"]) if sso_user else None
+    product_listings = ua_contracts_api.get_product_listings("canonical-cube")[
+        "productListings"
+    ]
+
+    assertions = {}
+    enrollments = []
+    passed_courses = 0
+    if edx_user:
+        assertions = {
+            assertion["badgeclass"]: assertion
+            for assertion in badgr_api.get_assertions(
+                BADGR_ISSUER, edx_user["email"]
+            )["result"]
+        }
+
+        enrollments = [
+            enrollment["course_details"]["course_id"]
+            for enrollment in edx_api.get_enrollments(edx_user["username"])
+            if enrollment["is_active"]
+        ]
+
+    certified_badge = {}
+    if CERTIFIED_BADGE in assertions:
+        assertion = assertions.pop(CERTIFIED_BADGE)
+        if not assertion["revoked"]:
+            certified_badge["image"] = assertion["image"]
+            certified_badge["share_url"] = assertion["openBadgeId"]
+
+    courses = []
+    for product_list in product_listings:
+        attempts = []
+        edx_ids = next(
+            (
+                ids
+                for ids in product_list["externalIDs"]
+                if ids["origin"] == "EdX"
+            ),
+            None,
+        )
+        if edx_ids:
```

Comment:
Slightly confused by this, is there any listing that has more than one EDX id?

Reply:
Well that's why I wanted you to take a look at this PR. I get a list of IDs, although currently I just get one in the list. Should I just grab the first always `edx_ids[0]`?

Reply:
I believe so, if a product listing represents a single EDX course it should only have one EDX ID

Reply:
`edx_id = [edx_id for edx_id in product_list["externalIDs"] if ids["origin"] == "EdX"][0]`

Reply:
I asked CS team what do they do with the IDs and they said that they just synch Edx with UA contracts using the email. So I trust that they always will give at least one ID?